buildSrc/src/main/java/org/elasticsearch/gradle/util/GradleUtils.java

```diff
@@ -120,4 +132,61 @@ public static void maybeConfigure(TaskContainer tasks, String name, Action<? sup
 
         return (Provider<T>) registration.getService();
     }
+
+    /**
+     * Add a source set and task of the same name that runs tests.
+     *
+     * IDEs are also configured if setup, and the test task is added to check. The new test source
+     * set extends from the normal test source set to allow sharing of utilities.
+     */
+    public static void addTestSourceSet(Project project, String sourceSetName) {
+        project.getPluginManager().apply(ElasticsearchJavaPlugin.class);
+
+        // create our test source set and task
+        SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
+        SourceSet extraTestSourceSet = sourceSets.create(sourceSetName);
+        TaskProvider<Test> testTask = project.getTasks().register(sourceSetName, Test.class);
+        testTask.configure(task -> {
+            task.setTestClassesDirs(extraTestSourceSet.getOutput().getClassesDirs());
+            task.setClasspath(extraTestSourceSet.getRuntimeClasspath());
+        });
+        SourceSet mainSourceSet = sourceSets.getByName("main");
+        SourceSet testSourceSet = sourceSets.getByName("test");
+
+        extendConfiguration(project, testSourceSet, extraTestSourceSet, SourceSet::getCompileConfigurationName);
```

Comment:
If I understand this (which I may not :) ) this applies the test configuration to the new source set ? (e.g. if testCompile depends on foo, then so does myNewTest sourceSet ?) 

Looking at this for the requirements for YAML tests, I wonder if this should be optional since we want the YAML tests to have no dependencies beyond the test framework.

Reply:
We could have this utility method accept a list of source sets from which the new source set should extend.

Reply:
I can work on that, although I'd like to do it in a followup. At first I tried having the tests only extend main. However, we have several IT tests in server which use static utility methods from unit test classes. It will take some work to sort that all out.